src/hotspot/share/gc/g1/g1ParScanThreadState.cpp

```diff
@@ -75,10 +76,17 @@ G1ParScanThreadState::G1ParScanThreadState(G1CollectedHeap* g1h,
     _old_gen_is_full(false),
     _partial_objarray_chunk_size(ParGCArrayScanChunk),
     _partial_array_stepper(n_workers),
+    _string_klass_or_null(G1StringDedup::is_enabled()
+                          ? SystemDictionary::String_klass()
+                          : nullptr),
     _num_optional_regions(optional_cset_length),
     _numa(g1h->numa()),
     _obj_alloc_stat(NULL)
 {
+  // Verify klass comparison with _string_klass_or_null is sufficient
```

Comment:
Not sure if this is the correct place to check this property: it needs not only hold during evacuation but basically anywhere. Is it possible to change finality of a java class?


What would be the impact of java.lang.String not passing this check? From what I understand only that these subclasses of j.l.String are not deduplicated. This does not seem to be a breaking (but surprising) problem, only that dedup for all subclasses of j.l.String does not work.

I would probably move that check into initialization, printing a warning.

Reply:
I don't believe finality can be changed in a relevant way. (Not counting changing the source code and recompiling, or perhaps patching the jar file.)  If some project were to try to change String, I think they'd very quickly trip this assert and know something needed to be done. Depending on what the change was, it could effectively disable deduplication (for example, if String were changed to be abstract).

I don't see a good way to check this during initialization that doesn't put the check far away from the code that cares, which makes it much less helpful.  I thought about putting the assert at the point of use, but that seemd a little excessive.